### PR TITLE
test: parameter server integration test

### LINKED ISSUE #645: [TEST] Parameter Server Integration
### What do you want to test?
Full integration test for parameter server.

### Outline of test structure
Should use all components typically found in the parameter server and run without error, but not using launchpad. See the jax IPPO example for how the components are typically logically distributed between core nodes. If multiple tests are required to cover different distinct cases (e.g. value-based vs policy-gradient), then they should be included.

### Definition of done
Test(s) passing and including all typical components.

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Parameter server integration test
## Why?
Part of on going efforts to increase testing code coverage of callback redesigned Mava.
## Extra
Close #645

### CODE REVIEW

File: tests/jax/systems/ippo_system_test.py [hunk 1]
```diff
@@ -106,14 +105,8 @@ def network_factory(*args: Any, **kwargs: Any) -> Any:
     trainer_node.disable_run()
```
Comment (AsadJeewa) [inline tests/jax/systems/ippo_system_test.py:105]: As discussed, lets disable all nodes and call them manually so we have control over execution (for all integration tests)

File: tests/jax/integration/parameter_server_test.py [hunk 2]
```diff
@@ -1,89 +1,357 @@
...
+    return mock_system_multi_process()
+
+
+@pytest.fixture
+def test_system_mt() -> System:
+    """A multi thread built system"""
+    return mock_system_multi_thread()
+
+
+def test_parameter_server_single_process(test_system_sp: System) -> None:
```
Comment (AsadJeewa) [inline tests/jax/integration/parameter_server_test.py:52]: same comments apply for other tests (multithread and multiprocess)

File: tests/jax/integration/parameter_server_test.py [hunk 3]
```diff
@@ -1,89 +1,357 @@
...
+    assert param_without_net == {
+        "trainer_steps": jnp.zeros(1, dtype=jnp.int32),
+        "trainer_walltime": jnp.zeros(1, dtype=jnp.float32),
+        "evaluator_steps": jnp.zeros(1, dtype=jnp.int32),
+        "evaluator_episodes": jnp.zeros(1, dtype=jnp.int32),
+        "executor_episodes": jnp.zeros(1, dtype=jnp.int32),
+        "executor_steps": jnp.zeros(1, dtype=jnp.int32),
+    }
+
+    ############################get_parameters test#####################################
```
Comment (AsadJeewa) [inline tests/jax/integration/parameter_server_test.py:81]: as discussed, think we need to make this more high-level and not test these methods that were covered in the unit test for the parameter server

File: tests/jax/integration/parameter_server_test.py [hunk 4]
```diff
@@ -1,89 +1,357 @@
...
+        )
+
+    # set non existing param
+    param1: Dict[str, str] = {"wrong_param": "test"}
+    with pytest.raises(AssertionError):
+        assert parameter_server.set_parameters(param1)
+
+    ############################add_to_parameters_test#####################################
+    param2: Dict[str, Any] = {
+        "trainer_steps": jnp.array([1, 3]),
```
Comment (AsadJeewa) [inline tests/jax/integration/parameter_server_test.py:130]: Issue here is that manually updating the params messes with the terminators (which we may not end up using though)

**Review by AsadJeewa — CHANGES_REQUESTED** [review]
Thanks @OmaymaMahjoub 
Comments based on our meeting this morning.

**Comment by AsadJeewa** [comment]
Apologies for the GitHub mess up. I reverted to your last commit.

File: tests/jax/integration/parameter_server_test.py [hunk 5]
```diff
@@ -1,89 +1,243 @@
...
+        "evaluator"
+    ]
+    (parameter_server_node,) = test_system_mp._builder.store.program._program._groups[
+        "parameter_server"
+    ]
+
+    trainer_node.disable_run()
+    executor_node.disable_run()
+    evaluator_node.disable_run()
+    # parameter_server_node.disable_run()
```
Comment (DriesSmit) [inline tests/jax/integration/parameter_server_test.py:195]: Why don't we disable the parameter server as well?

Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:195]: Yeah because I don't want to call explicitely the methods within the parameter server I wanted the executor and the trainer once they are running they will call set_parameter...

Comment (AsadJeewa) [reply tests/jax/integration/parameter_server_test.py:195]: also, the parameter server is disabled for multithread

Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:195]: Oh I didn't notice I will try to disable it and see

Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py]: It works even when I desable the node I updated the code! Thanks for noticing

File: tests/jax/integration/parameter_server_test.py [hunk 6]
```diff
@@ -1,89 +1,87 @@
...
+
+    trainer_steps = parameter_server.get_parameters("trainer_steps")
+    executor_episodes = parameter_server.get_parameters("executor_episodes")
+    assert list(trainer_steps) == [1]  # trainer.step one time
+    assert list(executor_episodes) == [3]  # run episodes three times
+
+    network = parameter_server.get_parameters("networks-network_agent")
+    for network in network.values():
+        assert sorted(list(network.keys())) == ["b", "w"]
+        assert jnp.size(network["w"]) != 0  # network is updated
```
Comment (DriesSmit) [inline tests/jax/integration/parameter_server_test.py:78]: Does this not just test that the networks are initialised and not really test for updating?

**Review by DriesSmit — COMMENTED** [review]
Looks good, thanks @OmaymaMahjoub :slightly_smiling_face: Just see my one comment.

File: tests/jax/integration/parameter_server_test.py (hunk 6 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:78]: but aren't we run the episode and the trainer so it w ill be updated?

Comment (DriesSmit) [reply tests/jax/integration/parameter_server_test.py:78]: But we are not testing for updating? We are testing if it is not zero? It is not zero after initialisation as well. Or am I misunderstanding the test?

File: tests/jax/integration/parameter_server_test.py [hunk 7]
```diff
@@ -1,89 +1,87 @@
...
+        "trainer_walltime": jnp.zeros(1, dtype=jnp.float32),
+        "evaluator_steps": jnp.zeros(1, dtype=jnp.int32),
+        "evaluator_episodes": jnp.zeros(1, dtype=jnp.int32),
+        "executor_episodes": jnp.zeros(1, dtype=jnp.int32),
+        "executor_steps": jnp.zeros(1, dtype=jnp.int32),
+    }
 
-    parameter_server.add_to_parameters({"trainer_steps": np.ones(1, dtype=np.int32)})
-    assert parameter_server.get_parameters("trainer_steps")[0] == 2
+    assert not parameter_server.store.last_checkpoint_time
```
Comment (AsadJeewa) [inline tests/jax/integration/parameter_server_test.py:57]: Maybe add a comment to explain that the checkpoint has not been saved yet

File: tests/jax/integration/parameter_server_test.py (hunk 6 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:78]: oh I see that make sense my bad I will make previous network and see if it is updated

Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py]: Done I updated the test

File: tests/jax/integration/parameter_server_test.py [hunk 8]
```diff
@@ -72,11 +68,17 @@ def test_parameter_server_single_process(test_system_sp: System) -> None:
...
-        assert jnp.size(network["w"]) != 0  # network is updated
-        assert jnp.size(network["b"]) != 0
+    # check that the network is updated (at least one of the values updated)
+    updated_network = parameter_server.get_parameters("networks-network_agent")
+    at_least_one_changed = False
+    for key in updated_network.keys():
+        assert sorted(list(updated_network[key].keys())) == ["b", "w"]
+        if not jnp.array_equal(
+            updated_network[key]["w"], first_network[key]["w"]
+        ) or not jnp.array_equal(updated_network[key]["b"], first_network[key]["w"]):
```
Comment (AsadJeewa) [inline tests/jax/integration/parameter_server_test.py:78]: first_network[key]["b"]

**Review by AsadJeewa — CHANGES_REQUESTED** [review]
Small changes requested @OmaymaMahjoub

File: tests/jax/integration/parameter_server_test.py [hunk 9]
```diff
@@ -1,89 +1,89 @@
...
-    parameter_server.add_to_parameters({"trainer_steps": np.ones(1, dtype=np.int32)})
-    assert parameter_server.get_parameters("trainer_steps")[0] == 2
+    trainer_steps = parameter_server.get_parameters("trainer_steps")
+    executor_episodes = parameter_server.get_parameters("executor_episodes")
+    assert list(trainer_steps) == [1]  # trainer.step one time
+    assert list(executor_episodes) == [3]  # run episodes three times
 
-    # Step the parameter sever
+    # check that the network is updated (at least one of the values updated)
+    updated_network = parameter_server.get_parameters("networks-network_agent")
```
Comment (AsadJeewa) [inline tests/jax/integration/parameter_server_test.py:72]: Also, this could be renamed to updated_networks_param

File: tests/jax/integration/parameter_server_test.py (hunk 8 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:78]: this is the old network before it gets updated it has the same keys as the updated_network

File: tests/jax/integration/parameter_server_test.py (hunk 9 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:72]: okay!

Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py]: Updated!

File: tests/jax/integration/parameter_server_test.py (hunk 7 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/parameter_server_test.py:58]: Done!

File: tests/jax/integration/parameter_server_test.py (hunk 9 above)
Comment (AsadJeewa) [reply tests/jax/integration/parameter_server_test.py]: same for first_network_param

File: tests/jax/integration/parameter_server_test.py (hunk 6 above)
Comment (DriesSmit) [reply tests/jax/integration/parameter_server_test.py]: Great, thanks! :slightly_smiling_face:

File: tests/jax/integration/parameter_server_test.py (hunk 8 above)
Comment (DriesSmit) [reply tests/jax/integration/parameter_server_test.py]: I think he is referring to the typo on line 78. It should be `first_network[key]["b"]`.

Comment (AsadJeewa) [reply tests/jax/integration/parameter_server_test.py]: yeah but we are comparing biases to weights here 
or not jnp.array_equal(updated_network_param[key]["b"], **first_network_param[key]["b"])**:

File: tests/jax/integration/parameter_server_test.py (hunk 9 above)
Comment (AsadJeewa) [reply tests/jax/integration/parameter_server_test.py]: Thanks

File: tests/jax/integration/parameter_server_test.py (hunk 8 above)
Comment (AsadJeewa) [reply tests/jax/integration/parameter_server_test.py]: Thanks

**Review by AsadJeewa — APPROVED** [review]
Thanks @OmaymaMahjoub

**Review by DriesSmit — APPROVED** [review]
Thanks @OmaymaMahjoub :slightly_smiling_face:
